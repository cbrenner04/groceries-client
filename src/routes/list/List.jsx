import React from 'react';
import Async from 'react-async';
import { useNavigate, useParams } from 'react-router-dom';

import { fetchList } from './utils';
import ListContainer from './containers/ListContainer';
import Loading from '../../components/Loading';
import UnknownError from '../error_pages/UnknownError';

function List() {
  const navigate = useNavigate();
  const { id } = useParams();

  return (
    <Async promiseFn={fetchList} id={id} navigate={navigate}>
      <Async.Pending>
        <Loading />
      </Async.Pending>
      <Async.Fulfilled>
        {(data) => (
          <ListContainer
            userId={data.currentUserId}
            list={data.list}
            purchasedItems={data.purchasedItems}
            categories={data.categories}
            listUsers={data.listUsers}
            includedCategories={data.includedCategories}
            notPurchasedItems={data.notPurchasedItems}
            permissions={data.permissions}
          />
        )}
      </Async.Fulfilled>
      <Async.Rejected>
        <UnknownError />
      </Async.Rejected>
    </Async>
  );
}

export default List;
