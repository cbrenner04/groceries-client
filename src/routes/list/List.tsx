import React from 'react';
import Async, { type PromiseFn } from 'react-async';
import { useNavigate, useParams } from 'react-router-dom';

import { fetchList } from './utils';
import ListContainer from './containers/ListContainer';
import Loading from '../../components/Loading';
import UnknownError from '../error_pages/UnknownError';
import type { IList, IListItem, IListUser } from '../../typings';

const List = (): React.JSX.Element => {
  const navigate = useNavigate();
  const { id } = useParams();

  // TODO: figure out `promiseFn` typings
  return (
    <Async promiseFn={fetchList as unknown as PromiseFn<void>} id={id} navigate={navigate}>
      <Async.Pending>
        <Loading />
      </Async.Pending>
      <Async.Fulfilled>
        {(data: {
          currentUserId: string;
          list: IList;
          purchasedItems: IListItem[];
          categories: string[];
          listUsers: IListUser[];
          includedCategories: string[];
          notPurchasedItems: Record<string, IListItem[]>;
          permissions: string;
        }): React.JSX.Element => (
          <ListContainer
            userId={data.currentUserId}
            list={data.list}
            purchasedItems={data.purchasedItems}
            categories={data.categories}
            listUsers={data.listUsers}
            includedCategories={data.includedCategories}
            notPurchasedItems={data.notPurchasedItems}
            permissions={data.permissions}
          />
        )}
      </Async.Fulfilled>
      <Async.Rejected>
        <UnknownError />
      </Async.Rejected>
    </Async>
  );
};

export default List;
