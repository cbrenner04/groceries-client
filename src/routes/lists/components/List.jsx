import React from 'react';
import PropTypes from 'prop-types';
import { Col, ListGroup, Row } from 'react-bootstrap';
import update from 'immutability-helper';
import { Link } from 'react-router-dom';

import { formatDate } from '../../../utils/format';
import listIconClass from '../../../utils/list_icon';
import { list } from '../../../types';

function List(props) {
  const handleListSelect = (list) => {
    const listIds = props.selectedLists.map((l) => l.id).join(',');
    let updatedLists;
    if (listIds.includes(list.id)) {
      updatedLists = props.selectedLists.filter((l) => l.id !== list.id);
    } else {
      updatedLists = update(props.selectedLists, { $push: [list] });
    }
    props.setSelectedLists(updatedLists);
  };

  const listName = (
    <h5 className="mb-1">
      <i className={`fa ${listIconClass(props.list.type)} text-secondary me-3`} />
      {props.listName}
    </h5>
  );

  const listNameElement = props.includeLinkToList ? (
    <Link to={`/lists/${props.list.id}`} className="router-link">
      {listName}
    </Link>
  ) : (
    listName
  );

  return (
    <ListGroup.Item
      className={props.listClass}
      style={{ display: 'block' }}
      data-test-class={props.testClass}
      data-test-id={`list-${props.list.id}`}
    >
      <Row className={props.multiSelect ? 'list-item-row' : ''}>
        {props.multiSelect && (
          <Col xs="1" className="mx-sm-auto">
            <input
              type="checkbox"
              style={{ position: 'absolute', top: '40%', left: '4%' }}
              onClick={() => handleListSelect(props.list)}
            />
            <div className="list-item-multi-divider"></div>
          </Col>
        )}
        <Col xs={props.multiSelect ? 10 : 12} sm={props.multiSelect ? 11 : 12}>
          <Row className={`${props.multiSelect ? 'ms-1' : ''} pt-1`}>
            <Col lg="6">{listNameElement}</Col>
            <Col lg="4" className={props.multiSelect ? 'list-multi-created-at' : ''}>
              <small className="text-muted">{formatDate(props.list.created_at)}</small>
            </Col>
            <Col lg="2" className={`${props.multiSelect ? 'list-multi-buttons' : ''} pe-lg-3`}>
              {props.listButtons}
            </Col>
          </Row>
        </Col>
      </Row>
    </ListGroup.Item>
  );
}

List.propTypes = {
  listButtons: PropTypes.element.isRequired,
  listName: PropTypes.string.isRequired,
  listClass: PropTypes.string.isRequired,
  testClass: PropTypes.string.isRequired,
  includeLinkToList: PropTypes.bool.isRequired,
  list: list.isRequired,
  multiSelect: PropTypes.bool.isRequired,
  selectedLists: PropTypes.arrayOf(list).isRequired,
  setSelectedLists: PropTypes.func.isRequired,
};

export default List;
