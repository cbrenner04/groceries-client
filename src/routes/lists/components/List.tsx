import React, { type ReactElement } from 'react';
import { Col, ListGroup, Row } from 'react-bootstrap';
import update from 'immutability-helper';
import { Link } from 'react-router-dom';

import { formatDate } from 'utils/format';
import listIconClass from 'utils/list_icon';
import type { IList } from 'typings';

export interface IListProps {
  listButtons: ReactElement;
  listName: string;
  listClass: string;
  testClass: string;
  includeLinkToList: boolean;
  list: IList;
  multiSelect: boolean;
  selectedLists: IList[];
  setSelectedLists: (lists: IList[]) => void;
}

const List: React.FC<IListProps> = (props): React.JSX.Element => {
  const handleListSelect = (list: IList): void => {
    const listIds = props.selectedLists.map((l) => l.id).join(',');
    let updatedLists;
    if (listIds.includes(list.id!)) {
      updatedLists = props.selectedLists.filter((l) => l.id !== list.id);
    } else {
      updatedLists = update(props.selectedLists, { $push: [list] });
    }
    props.setSelectedLists(updatedLists);
  };

  const listName = (
    <h5 className="mb-1">
      <i className={`fa ${listIconClass(props.list.type)} text-secondary me-3`} />
      {props.listName}
    </h5>
  );

  const listNameElement = props.includeLinkToList ? (
    <Link to={`/lists/${props.list.id}`} className="router-link">
      {listName}
    </Link>
  ) : (
    listName
  );

  return (
    <ListGroup.Item
      className={`${props.listClass} list-list-group-item`}
      data-test-class={props.testClass}
      data-test-id={`list-${props.list.id}`}
    >
      <Row className={props.multiSelect ? 'list-item-row' : ''}>
        {props.multiSelect && (
          <Col xs="1" className="mx-sm-auto">
            <input type="checkbox" className="multi-select-check" onClick={(): void => handleListSelect(props.list)} />
            <div className="list-item-multi-divider"></div>
          </Col>
        )}
        <Col xs={props.multiSelect ? 10 : 12} sm={props.multiSelect ? 11 : 12}>
          <Row className={`${props.multiSelect ? 'ms-1' : ''} pt-1`}>
            <Col lg="6">{listNameElement}</Col>
            <Col lg="4" className={props.multiSelect ? 'list-multi-created-at' : ''}>
              <small className="text-muted">{formatDate(props.list.created_at!)}</small>
            </Col>
            <Col lg="2" className={`${props.multiSelect ? 'list-multi-buttons' : ''} pe-lg-3`}>
              {props.listButtons}
            </Col>
          </Row>
        </Col>
      </Row>
    </ListGroup.Item>
  );
};

export default List;
