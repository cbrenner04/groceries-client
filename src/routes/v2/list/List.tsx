import React from 'react';
import Async, { type PromiseFn } from 'react-async';
import { useNavigate, useParams } from 'react-router';

import Loading from 'components/Loading';

import { fetchList, type IFulfilledListData } from './utils';
import ListContainer from './containers/ListContainer';
import UnknownError from '../../error_pages/UnknownError';

const List = (): React.JSX.Element => {
  const navigate = useNavigate();
  const { id } = useParams();

  return (
    <Async promiseFn={fetchList as unknown as PromiseFn<void>} id={id} navigate={navigate}>
      <Async.Pending>
        <Loading />
      </Async.Pending>
      <Async.Fulfilled>
        {(data: IFulfilledListData | undefined): React.JSX.Element => {
          // TODO: needed after updates to handleFailure - keep?
          if (!data) {
            return <UnknownError />;
          }
          return (
            <ListContainer
              userId={data.current_user_id}
              list={data.list}
              completedItems={data.completed_items}
              categories={data.categories}
              listUsers={data.list_users}
              notCompletedItems={data.not_completed_items}
              permissions={data.permissions}
              listsToUpdate={data.lists_to_update}
              listItemConfiguration={data.list_item_configuration}
            />
          );
        }}
      </Async.Fulfilled>
      <Async.Rejected>
        <UnknownError />
      </Async.Rejected>
    </Async>
  );
};

export default List;
