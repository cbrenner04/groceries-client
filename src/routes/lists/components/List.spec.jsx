import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import List from './List';

describe('List', () => {
  let props;
  const renderList = (p) => {
    return render(
      <MemoryRouter>
        <List {...p} />
      </MemoryRouter>,
    );
  };

  beforeEach(() => {
    props = {
      listButtons: <div />,
      listName: 'foo',
      listClass: 'foo',
      testClass: 'foo',
      includeLinkToList: false,
      list: {
        id: 'id1',
        name: 'foo',
        type: 'GroceryList',
        created_at: new Date('05/27/2020').toISOString(),
        completed: true,
        users_list_id: 'id1',
        owner_id: 'id1',
        refreshed: false,
      },
      multiSelect: false,
      selectedLists: [],
      setSelectedLists: jest.fn(),
    };
  });

  it('does not render Link when includeLinkToList is false', () => {
    props.includeLinkToList = false;
    const { container, queryByRole } = renderList(props);

    expect(container).toMatchSnapshot();
    expect(queryByRole('link')).toBeNull();
  });

  it('does render Link when includeLinkToList is true', () => {
    props.includeLinkToList = true;
    const { container, getByRole } = renderList(props);

    expect(container).toMatchSnapshot();
    expect(getByRole('link')).toBeVisible();
  });

  it('calls setSelectedLists with list when selecting list', async () => {
    props.multiSelect = true;
    props.setSelectedLists = jest.fn();

    const { container, getByRole } = renderList(props);

    expect(container).toMatchSnapshot();

    fireEvent.click(getByRole('checkbox'));

    expect(props.setSelectedLists).toHaveBeenCalledWith([props.list]);
  });

  it('calls setSelectedLists with empty array when deselecting list', async () => {
    props.multiSelect = true;
    props.setSelectedLists = jest.fn();
    props.selectedLists = [props.list];

    const { container, getByRole } = renderList(props);

    expect(container).toMatchSnapshot();

    fireEvent.click(getByRole('checkbox'));

    expect(props.setSelectedLists).toHaveBeenCalledWith([]);
  });
});
