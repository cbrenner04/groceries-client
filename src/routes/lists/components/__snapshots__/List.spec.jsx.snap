// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List calls setSelectedLists with empty array when deselecting list 1`] = `
<div>
  <div
    class="foo list-group-item"
    data-test-class="foo"
    data-test-id="list-id1"
    style="display: block;"
  >
    <div
      class="list-item-row row"
    >
      <div
        class="mx-sm-auto col-1"
      >
        <input
          style="position: absolute; top: 40%; left: 4%;"
          type="checkbox"
        />
        <div
          class="list-item-multi-divider"
        />
      </div>
      <div
        class="col-sm-11 col-10"
      >
        <div
          class="ms-1 pt-1 row"
        >
          <div
            class="col-lg-6"
          >
            <h5
              class="mb-1"
            >
              <i
                class="fa fa-shopping-bag text-secondary me-3"
              />
              foo
            </h5>
          </div>
          <div
            class="list-multi-created-at col-lg-4"
          >
            <small
              class="text-muted"
            >
              May 27 2020, 12:00:00 am
            </small>
          </div>
          <div
            class="list-multi-buttons pe-lg-3 col-lg-2"
          >
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`List calls setSelectedLists with list when selecting list 1`] = `
<div>
  <div
    class="foo list-group-item"
    data-test-class="foo"
    data-test-id="list-id1"
    style="display: block;"
  >
    <div
      class="list-item-row row"
    >
      <div
        class="mx-sm-auto col-1"
      >
        <input
          style="position: absolute; top: 40%; left: 4%;"
          type="checkbox"
        />
        <div
          class="list-item-multi-divider"
        />
      </div>
      <div
        class="col-sm-11 col-10"
      >
        <div
          class="ms-1 pt-1 row"
        >
          <div
            class="col-lg-6"
          >
            <h5
              class="mb-1"
            >
              <i
                class="fa fa-shopping-bag text-secondary me-3"
              />
              foo
            </h5>
          </div>
          <div
            class="list-multi-created-at col-lg-4"
          >
            <small
              class="text-muted"
            >
              May 27 2020, 12:00:00 am
            </small>
          </div>
          <div
            class="list-multi-buttons pe-lg-3 col-lg-2"
          >
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`List does not render Link when includeLinkToList is false 1`] = `
<div>
  <div
    class="foo list-group-item"
    data-test-class="foo"
    data-test-id="list-id1"
    style="display: block;"
  >
    <div
      class="row"
    >
      <div
        class="col-sm-12 col-12"
      >
        <div
          class=" pt-1 row"
        >
          <div
            class="col-lg-6"
          >
            <h5
              class="mb-1"
            >
              <i
                class="fa fa-shopping-bag text-secondary me-3"
              />
              foo
            </h5>
          </div>
          <div
            class="col-lg-4"
          >
            <small
              class="text-muted"
            >
              May 27 2020, 12:00:00 am
            </small>
          </div>
          <div
            class=" pe-lg-3 col-lg-2"
          >
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`List does render Link when includeLinkToList is true 1`] = `
<div>
  <div
    class="foo list-group-item"
    data-test-class="foo"
    data-test-id="list-id1"
    style="display: block;"
  >
    <div
      class="row"
    >
      <div
        class="col-sm-12 col-12"
      >
        <div
          class=" pt-1 row"
        >
          <div
            class="col-lg-6"
          >
            <a
              class="router-link"
              href="/lists/id1"
            >
              <h5
                class="mb-1"
              >
                <i
                  class="fa fa-shopping-bag text-secondary me-3"
                />
                foo
              </h5>
            </a>
          </div>
          <div
            class="col-lg-4"
          >
            <small
              class="text-muted"
            >
              May 27 2020, 12:00:00 am
            </small>
          </div>
          <div
            class=" pe-lg-3 col-lg-2"
          >
            <div />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
